File: src/cli/commands/record.ts
import { mkdirSync, writeFileSync } from 'node:fs';
import { join } from 'node:path';
import { genericAdapter } from '../../adapters/generic-adapter.js';
import { EventValidationError } from '../../core/event-schema.js';
import { fail, ok } from '../format.js';
import type { CommandResult } from '../types.js';

export interface RecordOptions {
  json: boolean;
}

export const STATE_DIR = '.masterytrace';
export const EVENTS_STATE_FILENAME = 'events.json';

/**
 * Validates and loads an event log (JSON or CSV, chosen by file extension)
 * and stores it to `.masterytrace/events.json`. Storing always *replaces*
 * any previously stored log in v0.1 (there is no append/merge mode yet).
 */
export function runRecord(cwd: string, eventLogPath: string, options: RecordOptions): CommandResult {
  let events;
  try {
    events = genericAdapter.load(eventLogPath);
  } catch (error) {
    if (error instanceof EventValidationError) {
      return fail(
        2,
        options.json,
        { error: error.message, issues: error.issues },
        `Validation error:\n${error.message}\n`,
      );
    }
    const message = error instanceof Error ? error.message : String(error);
    return fail(1, options.json, { error: message }, `Error: ${message}\n`);
  }

  const stateDir = join(cwd, STATE_DIR);
  mkdirSync(stateDir, { recursive: true });
  const statePath = join(stateDir, EVENTS_STATE_FILENAME);
  writeFileSync(statePath, `${JSON.stringify(events, null, 2)}\n`, 'utf-8');

  return ok(
    options.json,
    { eventCount: events.length, storedAt: statePath },
    `Stored ${events.length} event(s) to ${statePath}\n` +
      '(record replaces any previously stored event log; see --help for details.)\n',
  );
}

File: python/src/masterytrace/cli/commands/record.py
"""
Validates and loads an event log (JSON or CSV, chosen by file extension)
and stores it to `.masterytrace/events.json`. Storing always *replaces*
any previously stored log in v0.1 (there is no append/merge mode yet).
Ported from src/cli/commands/record.ts.
"""
from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path

from ...adapters.generic_adapter import generic_adapter
from ...core.event_schema import EventValidationError
from ..format import fail, ok
from ..json_encode import to_jsonable
from ..types import CommandResult

STATE_DIR = ".masterytrace"
EVENTS_STATE_FILENAME = "events.json"


@dataclass
class RecordOptions:
    json: bool


def run_record(cwd: str, event_log_path: str, options: RecordOptions) -> CommandResult:
    try:
        events = generic_adapter.load(event_log_path)
    except EventValidationError as error:
        return fail(
            2,
            options.json,
            {"error": str(error), "issues": error.issues},
            f"Validation error:\n{error}\n",
        )
    except Exception as error:  # noqa: BLE001 -- mirrors the TS catch-all for I/O/parse errors
        return fail(1, options.json, {"error": str(error)}, f"Error: {error}\n")

    state_dir = Path(cwd) / STATE_DIR
    state_dir.mkdir(parents=True, exist_ok=True)
    state_path = state_dir / EVENTS_STATE_FILENAME
    state_path.write_text(json.dumps(to_jsonable(events), indent=2) + "\n", encoding="utf-8")

    return ok(
        options.json,
        {"eventCount": len(events), "storedAt": str(state_path)},
        f"Stored {len(events)} event(s) to {state_path}\n"
        "(record replaces any previously stored event log; see --help for details.)\n",
    )
